### DISCUSSION TITLE
How to access '$this' in rules?

### DISCUSSION THREAD

**Discussion by francoism90** [spatie/laravel-data#150] [Q&A] [answered]
I know you can use the Unique rule, however I would like to use it on Livewire:
```
'post.name' => ['required', 'string', 'max:40', 'unique:'.Post::class.',name,'.$this->id],
```
How can one call $this->id?

**Comment by juse-less** [comment, marked as answer]
You should be able to access the `$id` property, by adding `array $payload` to your `rules()` method.

I didn't check the documentation right now, but immediately recalled PR #123.

```php
public static function rules(array $payload): array
{
    return [
        'post.name' => ['required', 'string', 'max:40', 'unique:'.Post::class.',name,'.$payload['id']],
    ];
}
```

**Reply by francoism90** [reply]
@juse-less Thanks! :)

**Reply by francoism90** [reply]
Hmm, this seems to be removed in V2.

Edit: seems `payload` needs to be called.

**Reply by alexrififi** [reply]
No. It remained in the V2.

**Reply by juse-less** [reply]
The call to the `Data::rules()` method is present [here](https://github.com/spatie/laravel-data/blob/main/src/Resolvers/DataClassValidationRulesResolver.php#L24-L28), in v2.

It's simply just passing on the `$payload` that is inserted to the `Data::validate()` method.
The only modification it does in the call chain (from what I can tell) is that it's casting `$payload`s implementing `\Illuminate\Contracts\Support\Arrayable` to an array.
Other than that, it doesn't look like it does more than just passing it around. 🙂
So, if it's empty it could be something else.
v1 required the param to be called [`$payload`](https://github.com/spatie/laravel-data/blob/v1/src/Resolvers/DataValidationRulesResolver.php#L22-L24) as well, so I reckon this isn't the issue here (wrong name).

**Comment by thewebartisan7** [comment]
And how you do this with `#Rule` attribute ?

```
    #[Rule('required', 'email', 'unique:users,email')]
    public ?string $email = null;
```

**Reply by rubenvanassche** [reply]
Use the `rules` method as described above with the `$payload` parameter.

**Reply by thewebartisan7** [reply]
yes thanks I see that it's not possible otherwise in another issue

**Reply by thewebartisan7** [reply]
Actually I use $request and works fine:

```php
public static function rules(Request $request): array
    {
        $unique = Rule::unique('users');
            $unique->ignore($request->user);
// ...

```

I think it's the same right?

**Reply by rubenvanassche** [reply]
You can inject anything in there so if the request is handy then that's a great way to solve this issue!
